nixos/coturn: set up sandboxing

mweinelt opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

Coturn is a network-facing implementation of the turn and stun protocols. It is a glorified udp/tcp relay at best and shall be restricted to that surface area.

```
default:   NAME                                                        DESCRIPTION                                                                                         EXPOSURE
✗ SystemCallFilter=~@resources                                System call allow list defined for service, and @resources is included (e.g. ioprio_set is allowed)      0.2
✗ SystemCallFilter=~@privileged                               System call allow list defined for service, and @privileged is included (e.g. setgroups is allowed)      0.2
✗ RootDirectory=/RootImage=                                   Service runs within the host's root directory                                                            0.1
✗ RestrictAddressFamilies=~AF_NETLINK                         Service may allocate netlink sockets                                                                     0.1
✗ RestrictAddressFamilies=~AF_(INET|INET6)                    Service may allocate Internet sockets                                                                    0.3
✗ PrivateNetwork=                                             Service has access to the host's network                                                                 0.5
✗ DeviceAllow=                                                Service has a device ACL with some special devices: char-rtc:r                                           0.1
✗ IPAddressDeny=                                              Service does not define an IP address allow list                                                         0.2

→ Overall exposure level for coturn.service: 1.3 OK :-)
```

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

mweinelt committed:
nixos/coturn: set up sandboxing

mweinelt committed:
nixos/coturn: reindent, unclutter

Make the module slightly easier to browse.

networkException commented on nixos/modules/services/networking/coturn.nix:
```diff
@@ -341,25 +341,66 @@ in {
+            "AF_INET"
+            "AF_INET6"
+            "AF_NETLINK"
+            "AF_UNSPEC"
```
I've not had to use this address family for sandboxing yet, is this something that the eBPF filter would hit?

mweinelt replied:
You can us AF_UNSPEC e.g. for IP family agnostic sockets, but I didn't check whether coturn actually only initializes the AF like that, or if it consistently overwrites it.

https://github.com/search?q=repo%3Acoturn%2Fcoturn%20AF_UNSPEC&type=code

networkException reviewed (commented):
Some thoughts / suggestions

mweinelt commented on nixos/modules/services/networking/coturn.nix:
```diff
@@ -341,25 +341,66 @@ in {
           chmod 640 ${runConfig}
         '';
-        serviceConfig = {
+        serviceConfig = rec {
```
Thanks, I created that list for synapse based on https://bgpfilterguide.nlnog.net/guides/bogon_prefixes/. Was thinking about including it, but that would make it hard to override, unless we make it an option.

That being said, coturn supports a `denied-peer-ip` setting, but it wants a non-cidr `<from>-<to>` format from what I remember.

```nix
    extraConfig = ''
      denied-peer-ip=10.0.0.0-10.255.255.255
      denied-peer-ip=192.168.0.0-192.168.255.255
      denied-peer-ip=172.16.0.0-172.31.255.255
    '';
```

networkException replied:
`IPAddressAllow` always takes precedence over `IPAddressDeny`, so overriding would be somewhat trivial. That said I'm very much not sure about how this would affect deployments

mweinelt replied:
I generally try not to make people override the systemd unit, as that is not well disoverable for new users.

mweinelt committed:
coturn: make setgroups conditional on privdrop codepath

Make coturn only call setgroups, when it actually needs to privdrop. In
the nixos module we already run coturn as an unprivileged user, which
means we don't need to provide access to the setgroups syscall in the
first place.

mweinelt commented on nixos/modules/services/networking/coturn.nix:
```diff
@@ -341,25 +341,66 @@ in {
+          SystemCallFilter = [
+            "@system-service"
+            "~@privileged"
+            "setgroups"
```
Very cool, thank you. Applied

mweinelt commented on nixos/modules/services/networking/coturn.nix:
```diff
@@ -341,25 +341,66 @@ in {
+          RestrictAddressFamilies = [
+            "AF_INET"
+            "AF_INET6"
+            "AF_NETLINK"
```
It's being called unconditionally on non-windows systems.

https://github.com/coturn/coturn/blob/08480ce137c57f93a8a25c57b192444a37f9254e/src/apps/relay/mainrelay.c#L671

networkException replied:
Its only used in a code path for bind address auto discovery, at least setting `listening-ip` avoids this.

It should be fine to keep `AF_NETLINK` to enumerate interfaces but I guess it would be nice if glibc could use some syscall instead of netlink (iirc despite some old proposal for selinux, there's no way to restrain netlink access further than user permissions and maybe capabilities) :shrug:

mweinelt replied:
Created this conditional relationship in the latest update.

mweinelt commented on nixos/modules/services/networking/coturn.nix:
```diff
@@ -341,25 +341,66 @@ in {
+          RestrictRealtime = true;
+          RestrictSUIDSGID = true;
+          SystemCallArchitectures = "native";
+          SystemCallFilter = [
```
And indeed I could still remove `@resources`, sneaky bugger.

networkException replied:
The output of `systemd-analyze syscall-filter` shows which individual system calls and other groups `@system-service` resolves to. That includes `@resources` directly (I'm not exactly sure why I usually disallow `@mount` as well, maybe that got moved out of `@system-service`, maybe just a subset of calls is included in both or `systemd-analyze security` just reports different things since some version)

mweinelt commented:
Anything else?

reivilibre commented:
@mweinelt Did you try coturn after this PR?
Since updating to 24.11 it seems coturn has completely broken on my system. This seems like the only suspicious change since 24.05.
I am still looking into it, but I think this PR also breaks the log output of coturn...

mweinelt commented:
Ugh, I'm sorry if I broke anything here.

I think the lack of a proper log could also be due to the hardcoded `no-stdout-log`.

I tested a few element calls in different network scenarios and all seem to have worked without any turn usage on my end.

mweinelt commented:
Dropping `no-stdout-log`, `syslog` and adding `verbose` gives a more interesting log at least.

reivilibre commented:
At the very least I can confirm this PR breaks the logging; I reverted this PR on my server and immediately I notice the log is no longer silent.

For reference, here is the config I've been using since 24.05 (and earlier?):

```nix
  # enable coturn
  services.coturn = rec {
    enable = true;
    no-cli = true;
    no-tcp-relay = true;
    min-port = 49000;
    max-port = 50000;
    use-auth-secret = true;
    static-auth-secret-file = config.sops.secrets."synapse/coturn_shared_secret".path;
    realm = "turn.librepush.net";
    cert = "${config.security.acme.certs.${realm}.directory}/full.pem";
    pkey = "${config.security.acme.certs.${realm}.directory}/key.pem";
    extraConfig = ''
      # for debugging
      verbose
      # ban private IP ranges
      no-multicast-peers
      denied-peer-ip=0.0.0.0-0.255.255.255
      denied-peer-ip=10.0.0.0-10.255.255.255
      denied-peer-ip=100.64.0.0-100.127.255.255
      denied-peer-ip=127.0.0.0-127.255.255.255
      denied-peer-ip=169.254.0.0-169.254.255.255
      denied-peer-ip=172.16.0.0-172.31.255.255
      denied-peer-ip=192.0.0.0-192.0.0.255
      denied-peer-ip=192.0.2.0-192.0.2.255
      denied-peer-ip=192.88.99.0-192.88.99.255
      denied-peer-ip=192.168.0.0-192.168.255.255
      denied-peer-ip=198.18.0.0-198.19.255.255
      denied-peer-ip=198.51.100.0-198.51.100.255
      denied-peer-ip=203.0.113.0-203.0.113.255
      denied-peer-ip=240.0.0.0-255.255.255.255
      denied-peer-ip=::1
      denied-peer-ip=64:ff9b::-64:ff9b::ffff:ffff
      denied-peer-ip=::ffff:0.0.0.0-::ffff:255.255.255.255
      denied-peer-ip=100::-100::ffff:ffff:ffff:ffff
      denied-peer-ip=2001::-2001:1ff:ffff:ffff:ffff:ffff:ffff:ffff
      denied-peer-ip=2002::-2002:ffff:ffff:ffff:ffff:ffff:ffff:ffff
      denied-peer-ip=fc00::-fdff:ffff:ffff:ffff:ffff:ffff:ffff:ffff
      denied-peer-ip=fe80::-febf:ffff:ffff:ffff:ffff:ffff:ffff:ffff
    '';
  };
```

My suspicion is that some of the hardening options here are blocking the syslog entries (not sure why this service is set up to log via syslog instead of stdout)

Now voice calls have started working and also appear to work even with this PR applied again, so nyurrghghhh I don't know if there's any problem from this PR other than the logging. I'll keep poking and get the people who were having trouble to test some more

mweinelt commented:
Restoring logging in https://github.com/NixOS/nixpkgs/pull/369828.
